backbone.js

```diff
@@ -658,29 +679,33 @@
     destroy: function(options) {
       options = options ? _.clone(options) : {};
       var model = this;
-      var success = options.success;
       var wait = options.wait;
+      var success = options.success;
+      var error = options.error;
+      delete options.success;
+      delete options.error;
 
       var destroy = function() {
         model.stopListening();
         model.trigger('destroy', model, model.collection, options);
       };
 
-      options.success = function(resp) {
-        if (wait) destroy();
-        if (success) success.call(options.context, model, resp, options);
-        if (!model.isNew()) model.trigger('sync', model, resp, options);
-      };
-
-      var xhr = false;
-      if (this.isNew()) {
-        _.defer(options.success);
-      } else {
-        wrapError(this, options);
-        xhr = this.sync('delete', this, options);
-      }
+      var isNew = model.isNew();
+      var promise = isNew ? Promise.resolve(model.attributes) :
+        wrapError(this.sync('delete', this, options), this, options, error);
       if (!wait) destroy();
-      return xhr;
+      return promise.then(function(resp) {
+        var promise = Promise.resolve(model);
+        if (wait) destroy();
+        if (success)
+          promise = Promise.resolve(success.call(options.context, model, resp, options));
+        if (!isNew)
+          promise = promise.then(function(delivered) {
+            model.trigger('sync', model, resp, options);
```

Comment:
Sorry, I got confused, the error callback provides the hook to resolve a rejected promise from `sync` (not the success callback). The success callback provides a hook to change the object (by reference) with which the returned `Promise` is fulfilled (and/or modify the model, if it wants) before the sync event is triggered. Event#trigger doesn't return anything particularly useful to the chain of promises so this is the only chance that the destroy method can participant in the promise chain after the call to sync and before sync is triggered.

Thinking about it some more, you could argue that the sync event doesn't care how or what the promise is fulfilled (i.e. `delivered` is not passed to Events#trigger), so what does it matter...